### PR TITLE
Feature/pcr product as sequence rebased

### PR DESCRIPTION
**Pull request by AJamesPhillips**
There may be regression or a bad merge as annotations on plasmid map are not showing.

### CODE REVIEW

File: public/scripts/sequence/models/sequence.js [hunk 1]
```diff
@@ -1050,6 +1071,20 @@ SequenceModel.concatenateSequences = function(sequences, circularise=false, trun
 };
 
 
+SequenceModel.calculateSequence = function(sequenceNts, opts) {
```
Comment (depha) [inline public/scripts/sequence/models/sequence.js]: This name is a bit obscure. `calculateProduct` might be more obvious?

Comment (AJamesPhillips) [inline public/scripts/sequence/models/sequence.js]: Agreed.  Spot on.

File: public/scripts/plugins/pcr/lib/utils.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  return products;
+};
+
+
+var savePcrProductsToSequence = function(sequenceModel, products) {
+  // Originally the model attributes were just stored and handled as a hash,
+  // we now want to use a model to handle them.
+  // We call `stringify` then `parse`, to convert both vanilla hashes and
+  // backbone models into vanilla hashes.
+  var attributesOfPcrProducts = JSON.parse(JSON.stringify(products));
```
Comment (depha) [inline public/scripts/plugins/pcr/lib/utils.js]: there's got to be a more efficient of doing that.

How about something like 
```js
var attributesOfPcrProducts = (products.toJSON && products.toJSON) || products;
```

File: public/scripts/plugins/pcr/lib/sequence.js [hunk 3]
```diff
@@ -27,7 +27,11 @@ class Sequence {
   }
 
   allFields () {
-    return this.requiredFields().concat(['id', 'name', 'antisense']);
+    return this.requiredFields().concat(this.optionalFields());
+  }
+
+  optionalFields () {
```
Comment (depha) [inline public/scripts/plugins/pcr/lib/sequence.js]: why are `optionalFields` and `requiredFields` functions?

Comment (AJamesPhillips) [inline public/scripts/plugins/pcr/lib/sequence.js]: Unless we can declare them `readonly` I prefer to put these as functions that return a fresh copy of an array each time.  Things can get accidentally edited and it becomes very confusing.  For example `slice` will edit an array associated as an attribute.  Just defensive programming.

File: public/scripts/plugins/pcr/lib/utils.js (hunk 2 above)
Comment (AJamesPhillips) [inline public/scripts/plugins/pcr/lib/utils.js]: Yeah happy to change if you want.  I just tried it for long sequences (1e7) and it returned instantaneously, but `(products.toJSON && products.toJSON()) || products;` should work better.

**Pull request merged**